### ISSUE TITLE
Chapter 14, Results of creative new sequence are quite different.

### ISSUE THREAD

**Issue by tetsuyasu** [ageron/handson-ml#274]
Thank you for improving the codes.

I am examining the results of chapter 14.
But my results of creative new sequence (In[64], In[65], and In[84]) are quite different from yours.
And also my result of embedded reber grammers (In[129]) is different from yours.
I used the same seed=42 for random.

All three are different,
- Your results
- My results on Mac
- My results on Ubuntu

Is this OK?
Can you explain the difference by randomness of learning or optimizer?

---

Examples of differences:

In[64]:
Your result is going up from 0 to 14.
My result on Mac is going down from 0 to -3.7.
My result on Ubuntu is going down from 0 to -50.

In[65] right:
Your result vibrates between 10 and 30.
My result on Mac vibrates between -5 and -2.
My result on Ubuntu is going down from -10 to -70.

In[129]:
Your result is 0.00%/100.00% (absolute confidence).
My result on Mac is 38.38%/99.93%.
My result on Ubuntu is 0.43%/100.00%.

---

Versions:

Mac:
Python 3.5.5 :: Anaconda custom (64-bit)
tensorflow==1.9.0
numpy==1.14.5
mkl-random==1.0.1

Ubuntu:
Python 3.5.5 :: Anaconda, Inc.
tensorflow-gpu==1.9.0
numpy==1.14.5
mkl-random==1.0.1

**Comment by ageron** [comment]
Hi @tetsuyasu ,
Thanks for your feedback. That's odd...  Could you please tell me whether the result changes on your machines when you restart and run the notebook? I need to know whether there's simply a seed missing somewhere (which would explain why the result changes every time) or whether the result is stable on one machine, but different across platforms.
Thanks,
Aurélien

**Comment by tetsuyasu** [comment]
Thank you for your response.
I have compared my two results. (1st) restart and run, (2nd) restart and run.
But the comparison results are very complicated.

On Mac (without GPU):
In[64]: almost same
In[65]: almost same
In[84]:  almost same
In[129]: quite different

On Ubuntu (with GPU)
In[64]: quite different
In[65]: quite different
In[84]:  little different
In[129]: quite different

It is very strange.

**Comment by ageron** [comment]
Hi @tetsuyasu ,
Thanks for running these tests. I just ran the notebook and indeed I get different results than the version on github.  :(  Since I created this notebook, I updated my TensorFlow version (1.7 to 1.10-rc1) and my Python version (3.5 to 3.6).  At the beginning of the notebook, the differences are really tiny, so I'm guessing this is due to tiny implementation differences in TensorFlow or Python, or one of the libraries they rely on. However, I get very different results in the cells you mention (and others).  I'm running on my MacBook. I'll run the notebook a second time to see what's stable. This is so strange...

**Comment by tetsuyasu** [comment]
Thank you for understanding the situation.
If you need, I can run the code with changing Python or library version.

**Comment by ageron** [comment]
Okay I am running the notebook a second time. So far, everything is exactly identical to the previous run (but of course still different from the version on github), except:
* cells In[82] and In[84]. I suspect something fishy in the dropout implementation. I'll double-check this.
* cells In[102] and In[108]. This is certainly due to the use of the `Counter`. Since it is a dictionary, the order of its elements is not guaranteed to be stable across multiple runs. However, the differences are not huge because of the call to `most_common()` which sorts most words identically across runs, except the ones that appear the exact same amount of times. To fix this, I should sort the output of `most_common()` so that words with equal counts are sorted alphabetically, this way it should be stable across runs.
The notebook is still running 108, I'll let you know when it finishes.

**Comment by ageron** [comment]
Could you please install Python 3.6.6 and TensorFlow 1.10.0-rc1 and see if you get the same results across multiple runs (except cells 82, 84, 102 and 108)? Thanks for your kind help!

**Comment by tetsuyasu** [comment]
Ok, I will try as you say.

**Comment by ageron** [comment]
The reber grammar example was also different across runs (and different from the github version), but oddly it was perfectly repeatable if I just run the code again without restarting the kernel. I ended finding the bug in the `generate_corrupted_string()` function. It uses `set()` which hashes the elements you add to it to decide where to place them. For this it calls the built-in `hash()` function. This function is stable within one run, but not stable across runs, at least not for strings:

```python
>>> hash("a")
-7814606714743762510
>>> hash("a")
-7814606714743762510
```

After restarting the shell:

```python
>>> hash("a")
-6256206269043620852
>>> hash("a")
-6256206269043620852
```

This means that sets have a stable order within a given run, but not across runs:

```python
>>> set("abcd")
{'d', 'a', 'c', 'b'}
>>> set("abcd")
{'d', 'a', 'c', 'b'}
```

And after a restart:

```python
>>> set("abcd")
{'d', 'a', 'c', 'b'}
>>> set("abcd")
{'d', 'a', 'c', 'b'}
```

So in the `generate_corrupted_string()` function, the following line is not stable across runs:

```python
bad_char = choice(list(set(chars) - set(good_char)))
```

It's easy to fix: just replace `list` with `sorted`, so here's the updated function:

```python
def generate_corrupted_string(grammar, chars="BEPSTVX"):
    good_string = generate_string(grammar)
    index = np.random.randint(len(good_string))
    good_char = good_string[index]
    bad_char = choice(sorted(set(chars) - set(good_char)))
    return good_string[:index] + bad_char + good_string[index + 1:]
```

The bad news is that the final example now fails!

```
BPBTSSSSSSSSSSSSXXTTTTTVPXTTVPXTTTTTTTVPXVPXVPXTTTVVETE: 64.68%
BPBTSSSSSSSSSSSSXXTTTTTVPXTTVPXTTTTTTTVPXVPXVPXTTTVVEPE: 99.99%
```

Oh nooo... well, we could work on improving the model, but I'm a bit out of time so I'll just give it a simpler example by removing a few `S`:

```
BPBTSSSSSSSXXTTTTTVPXTTVPXTTTTTTTVPXVPXVPXTTTVVETE: 5.42%
BPBTSSSSSSSXXTTTTTVPXTTVPXTTTTTTTVPXVPXVPXTTTVVEPE: 100.00%
```

**Comment by ageron** [comment]
I pushed a new notebook which is mostly reproducible, except for the dropout section. I really can't make it reproducible, it's frustrating.  I suspect that setting the number of inter-op threads to 1 would help (if so, then it might be possible to get reproducible on a GPU, since it does not use inter-op parallelism, only intra-op).  I gotta go, so I'll try this another day. Cheers!

**Comment by tetsuyasu** [comment]
I have finished the 1st run.

---
Environment:

Mac:
Python 3.6.6 :: Anaconda custom (64-bit)
tensorflow==1.10.0rc1
numpy==1.14.5

Ubuntu:
Python 3.6.6 :: Anaconda custom (64-bit)
tensorflow-gpu==1.10.0rc1
numpy==1.14.5

---

Comparison:

yours and 1st run on Mac:
In[64]: same
In[65]: same
In[84]: little different
In[130]: same

yours and 1st run on Ubuntu:
In[64]: different
In[65]: different
In[84]: little different
In[130]: different (yours=13.30%/99.97%, mine=2.83%/99.98%)

---

**Comment by tetsuyasu** [comment]
I have finished the 2nd run.
It seems that there is no problem when not using GPU.
Is there anything else I can do?

---

Environment:

Mac:
Python 3.6.6 :: Anaconda custom (64-bit)
tensorflow==1.10.0rc1
numpy==1.14.5

Ubuntu:
Python 3.6.6 :: Anaconda custom (64-bit)
tensorflow-gpu==1.10.0rc1
numpy==1.14.5

---

Comparison:

1st and 2nd run on Mac:
In[64]: same
In[65]: same
In[84]: little different
In[130]: same

1st and 2nd run on Ubuntu:
In[64]: different
In[65]: left different, right almost same
In[84]: little different
In[130]: almost same (1st=2.83%/99.98%, 2nd=2.32%/99.93%)

---

**Comment by ageron** [comment]
Hi @tetsuyasu ,

Thanks so much for all these tests, it must have been very time consuming, I appreciate it.  I never tried to get perfectly identical results when working on a GPU, only when working on these notebooks on my MacBook (CPU). Unfortunately, after further investigation it seems that it's impossible to get reproducible results when using TensorFlow on a GPU. This Kaggle discussion explains it very well, in particular Marcin Pękalski's answer:

> Calculations in finite arithmetic depend on the order they are performed. If the order is non-deterministic so will be the results. Both TF and PyTorch use CUDNN, which by default is non-deterministic. In PyTorch you may set a flag that will use deterministic methods of CUDNN, this option does not exist in TF.
> 
> Another thing is number of threads, you should make dataset reader to be single-threaded.
> 
> You can check that if you switch to using only CPU the results should be deterministic, even with TF.

This is indeed what you have observed: deterministic results on the CPU (except for the annoying `DropoutWrapper` issue), but non-deterministic on the GPU.

When using PyTorch and forcing cuDNN to be deterministic, it seems that performance drops dramatically. For example, in [this issue](https://github.com/keras-team/keras/issues/2280) @diogoff wrote:

> I gave up on reproducibility because I found that when forcing deterministic behavior in cuDNN, training would be much slower (e.g. from 15 secs/epoch to 30 secs/epoch).

Similarly, when using the CPU, you may have to set the interop threads to 1 (effectively turning off multithreading) to get reproducible results. This will hinder performance a lot. 

So the question is: is it worth it? In my case, I want the notebooks to have stable outputs so that the examples in the book will have the same output when people try them. Unfortunately, the result will not be the same on a GPU, but I suspect that most readers learn on a CPU.

In your case, what is your motivation for reproducibility? I have worked with researchers who cared about reproducibility because they wanted to be able to share experiments with their colleagues and for them to obtain the same exact results.  Maybe a better approach to reproducibility in this case is to run the non-reproducible experiment 10 times and look at the stats: mean, std dev, etc. It should be fairly stable, and it will tell you more about the model.

I'll add a note about reproducibility in the next edition, thanks to you! :)

Cheers,
Aurélien

**Comment by ageron** [comment]
Note that TensorFlow does not rely on cuDNN for every operation, but only for complex and intensive computations such as convolutions and LSTMs. So if you build a simple graph without these operations, it may be reproducible on the GPU. Unfortunately, some core TensorFlow functions are non-reproducible on the GPU, for example `tf.reduce_sum()`. Even if you don't use this function directly, it may be used automatically by TensorFlow to compute the gradients. This is very well explained in this [fantastic post](https://www.twosigma.com/insights/article/a-workaround-for-non-determinism-in-tensorflow/) by Two Sigma on achieving reproducibility with TensorFlow on the GPU.

Even though they manage to get their results to be reproducible, it seems to me that it's way too much effort. If you absolutely need to get reproducible results, I would just recommend running the code on a CPU, turning multithreading off and setting the random seeds.

**Comment by tetsuyasu** [comment]
Hi Aurélien,

My motivations for reproducibility are as follows.

(1) I was worried whether only my environment is so strange.
-> Because you explained the mechanism, it does not matter.

(2) I am using your book as my student's textbook. Therefore, I'm worried about students would get confused if the results were significantly different.
-> I understand that we should run on CPU with a single thread.

I have no problem, so I will close this issue.
